Classification metrics overhaul: accuracy metrics (2/n)

tadejsv opened the pull request:
This PR is a spin-off from #4835.

## What does this PR do?

### Metrics (changes and additions)

The metrics below are based on the new input formatting function from #4837

#### Accuracy

The accuracy metric now get a new `subset_accuracy` parameter, to calculate subset accuracy in case of multi-label or multi-dimensional multi-class inputs. This parameter is set to `False` by default, which preserves the default behavior of the old metric.

Additionally, a `top_k` parameter enables it to be used as TopK accuracy for multi-class inputs (with probabilities) - thanks @rohitgr7 . The `top_k` format does not yet have an equivalent in sklearn (will in 0.24).

#### HammingDistance (new)

This is equivalent to `hamming_loss` from sklearn.

pep8speaks commented:
Hello @tadejsv! Thanks for opening this PR. 











* In the file [`pytorch_lightning/metrics/functional/accuracy.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/35627b50092df312ed71d9ab1c828a2451306d5d/pytorch_lightning/metrics/functional/accuracy.py):

> [Line 71:36](https://github.com/PyTorchLightning/pytorch-lightning/blob/35627b50092df312ed71d9ab1c828a2451306d5d/pytorch_lightning/metrics/functional/accuracy.py#L71): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace











Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

Borda commented:
@tadejsv the first one got merged, can we continue with this one? :]

tadejsv commented:
> @tadejsv the first one got merged, can we continue with this one? :]

Yes, ready for review!

@SkafteNicki @teddykoker @awaelchli @rohitgr7

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+        top_k:
+            Number of highest probability entries for each sample to convert to 1s, relevant
+            only for (multi-dimensional) multi-class inputs with probability predictions. The
+            default value (``None``) will be interpreted as 1 for these inputs.
```
Why is default not 1?

rohitgr7 replied:
this handles multi-label and binary cases too, so `top_k=1` won't make sense there. Ideally, it should raise an error in such a case. haven't checked yet. if it does.

tadejsv replied:
It does raise an error for such cases - through the formatting function from the previous PR :) And yes, I agree with this - making it default to 1 could be potentially misleading in cases where top_k does not really apply.

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+            ``N`` dimension - that is, for the sample to count as correct, all labels on its extra dimension
+            must be predicted correctly (the ``top_k`` option still applies here). The final score is then
+            simply the number of totally correctly predicted samples.
+        top_k:
```
Can we exchange the order of `top_k` and `mdmc_accuracy`? I feel more users will be using `top_k` argument.

tadejsv replied:
Will do

SkafteNicki commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -22,9 +22,10 @@
-        pred: torch.Tensor,
-        target: torch.Tensor,
-        class_index: int, argmax_dim: int = 1,
+    pred: torch.Tensor,
```
Could you remove all formatting changes?
I assume that what is actually changing in this file is that you are removing the old accuracy implementation?

tadejsv replied:
Sure. These formatting changes appeared bc I automatically formatted it with black 😅

SkafteNicki commented on tests/metrics/functional/test_classification.py:
```diff
@@ -27,16 +25,18 @@
 
 
-@pytest.mark.parametrize(['sklearn_metric', 'torch_metric', 'only_binary'], [
-    pytest.param(sk_accuracy, accuracy, False, id='accuracy'),
```
This is the only line that really is changing right? (removing accuracy from tests).
If so, please remove all the other formatting changes

SkafteNicki reviewed (commented):
PR is looking very good, great job :]

Borda commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
         threshold:
-            Threshold value for binary or multi-label logits. default: 0.5
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            `(0,1)` predictions, in the case of binary or multi-label inputs. Default: `0.5`
```

Borda commented on pytorch_lightning/metrics/classification/hamming_loss.py:
```diff
@@ -0,0 +1,104 @@
+    Args:
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            `(0,1)` predictions, in the case of binary or multi-label inputs. Default: `0.5`
```

Borda commented on pytorch_lightning/metrics/classification/hamming_loss.py:
```diff
@@ -0,0 +1,104 @@
+
+        self.add_state("correct", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
+
```
```suggestion

        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

Borda commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -83,23 +111,28 @@ def __init__(
         self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
 
```
```suggestion

        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+
+            Should be left at default (``None``) for all other types of inputs.
+        mdmc_accuracy:
+            Determines how should the extra dimension be handeled in case of multi-dimensional multi-class
```
```suggestion
            Determines how should the extra dimension be handled in case of multi-dimensional multi-class
```

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+            If ``"global"``, then the inputs are treated as if the sample (``N``) and the extra dimension
+            were unrolled into a new sample dimension.
+
+            If ``"subset"``, than the equivalent of subset accuracy is performed for each sample on the
```
```suggestion
            If ``"subset"``, then the equivalent of subset accuracy is performed for each sample on the
```

Borda commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,114 @@
+        >>> accuracy(preds, target, top_k=2)
+        tensor(0.6667)
+    """
+
```
```suggestion
        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

tadejsv replied:
Not really needed here since the input formatting function raises a value error for this exact thing

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
 
-    - ``preds`` (float or long tensor): ``(N, ...)`` or ``(N, C, ...)`` where C is the number of classes
-    - ``target`` (long tensor): ``(N, ...)``
+    This metric generalizes to subset accuracy for multilabel data: for the sample to be counted as
+    correct, all labels in that sample have to be correctly predicted. Consider using :class:`~pytorch_lightning.metrics.classification.HammingLoss`
+    is this is not what you want. In multi-dimensional multi-class case the `mdmc_accuracy` parameters
+    gives you a choice between computing the subset accuracy, or counting each sample on the extra
+    axis separately.
```
```suggestion
    This metric generalizes to subset accuracy for multilabel data: for the sample to be counted as
    correct, all labels in that sample have to be correctly predicted. Consider using :class:`~pytorch_lightning.metrics.classification.HammingLoss`
    is this is not what you want. In a multi-dimensional multi-class case, the `mdmc_accuracy` parameters
    gives you a choice between computing the subset accuracy or counting each sample on the extra
    axis separately.
```

Borda commented on pytorch_lightning/metrics/functional/hamming_loss.py:
```diff
@@ -0,0 +1,68 @@
+        tensor(0.2500)
+
+    """
+
```
```suggestion
        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

Borda reviewed (commented):
let's add a few more assets (or maybe better raise `ValueError`) for arguments which are expected only in a given range or a limited set of values...

rohitgr7 commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,114 @@
+
+            Should be left at default (``None``) for all other types of inputs.
+        mdmc_accuracy:
+            Determines how should the extra dimension be handeled in case of multi-dimensional multi-class
```
```suggestion
            Determines how should the extra dimension be handled in case of multi-dimensional multi-class
```

rohitgr7 commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,114 @@
+            If ``"global"``, then the inputs are treated as if the sample (``N``) and the extra dimension
+            were unrolled into a new sample dimension.
+
+            If ``"subset"``, than the equivalent of subset accuracy is performed for each sample on the
```
```suggestion
            If ``"subset"``, then the equivalent of subset accuracy is performed for each sample on the
```

rohitgr7 commented on pytorch_lightning/metrics/functional/hamming_loss.py:
```diff
@@ -0,0 +1,68 @@
+        tensor(0.2500)
+
+    """
+        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```
```suggestion
    assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

rohitgr7 commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,114 @@
+        >>> accuracy(preds, target, top_k=2)
+        tensor(0.6667)
+    """
+        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```
```suggestion
    assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -63,10 +82,19 @@ class Accuracy(Metric):
         self,
         threshold: float = 0.5,
+        top_k: Optional[int] = None,
+        mdmc_accuracy: str = "subset",
```
To be backward compatible the default should be `global`:
```suggestion
        mdmc_accuracy: str = "global",
```

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+            default value (``None``) will be interpreted as 1 for these inputs.
+
+            Should be left at default (``None``) for all other types of inputs.
+        mdmc_accuracy:
```
I wonder if we can come up with another name for this parameter, that is a bit more descriptive?

tadejsv replied:
Hm, how about if we replace this parameter with one called `subset_accuracy`, and set it to `False` by default? 

This would then have an effect in the case of multi-label inputs (`=True` means the whole sample has to be correct, `=False` would count each label separately), and for mdmc inputs (`=True` equivalent to `subset` now and `=False` equivalent to `global` now)?

tadejsv replied:
Ok, actually I think this is the perfect solution - enables us to keep the default behavior of the old accuracy for all cases, while giving users the option to compute subset_accuracy for ml and mdmc inputs.

I went ahead and implemented it, but let me know if you have some objections/other ideas.

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,52 @@
+            inputs. Options are ``"global"`` or ``"subset"``.
+
+            If ``"global"``, then the inputs are treated as if the sample (``N``) and the extra dimension
+            were unrolled into a new sample dimension.
```
Can you explicit incorporate that `global` = internally flattens arrays

tadejsv replied:
What do you mean by this?

SkafteNicki replied:
If `mdmc_accuracy`=`global` give the same result as calling the metric on `accuracy(preds.flatten(), target.flatten())`

tadejsv replied:
Well yes - if your inputs are multi-dimensional multi-class with labels. But this is not true in general - for multi-label inputs this parameter does nothing (subset accuracy is always computed for them), and of course for multi-dimensional multi-class inputs with probability predictions this can't be the case as inputs need to be first transformed into the binarized form.

tadejsv replied:
But yeah, I can add this to the description, it should make it clearer

SkafteNicki replied:
Only a minor comment, It is only because this was the behavior of the old accuracy implementation

tadejsv replied:
Alright, I updated the description now

SkafteNicki commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -82,24 +110,36 @@ def __init__(
+            raise ValueError("The `threshold` should lie in the [0,1] interval.")
+
         self.threshold = threshold
+        self.top_k = top_k
```
Check that `top_k` is int larger than 0 (if defined)?

SkafteNicki commented on pytorch_lightning/metrics/classification/hamming_loss.py:
```diff
@@ -0,0 +1,105 @@
+        self.add_state("correct", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
+
+        assert 0 <= threshold <= 1, f"threshold: {threshold} is out of range"
```
Change to a `raise ValueError` instead (similar to accuracy)

SkafteNicki commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,117 @@
+        correct = (sample_correct == sample_total).sum()
+        total = target.shape[0]
+
+    return (torch.tensor(correct, device=preds.device), torch.tensor(total, device=preds.device))
```
`correct` should always be on the correct device (I think), so no need to do `torch.tensor` on that.
For `total` only two of the cases needs casting (first and last). Please convert in these two cases such that the return statement simply becomes:
```suggestion
    return correct, total
```

SkafteNicki commented on tests/metrics/classification/test_accuracy.py:
```diff
@@ -13,101 +15,129 @@
-    sk_target = target.view(-1).numpy()
 
-    return accuracy_score(y_true=sk_target, y_pred=sk_preds)
+# Replace with a proper sk_metric test once sklearn 0.24 hits :)
```
Very nice :]

Borda commented on pytorch_lightning/metrics/functional/hamming_loss.py:
```diff
@@ -0,0 +1,68 @@
+    return 1 - correct.float() / total
+
+
+def hamming_loss(preds: torch.Tensor, target: torch.Tensor, threshold: float = 0.5) -> torch.Tensor:
```
shall be loss separate from metrics?

tadejsv replied:
So this is called a loss only because better prediction result in lower values of the metric. It is not a loss in a sense that this is anything you could (or would want to) use to train your model. It's kind of like 1-accuracy.

tadejsv replied:
We could also name it Hamming distance, but I chose Hamming loss to stay in line with sklearn's naming

justusschock replied:
@tadejsv I would name it distance to avoid confusion here...

tadejsv replied:
Alright, I'll rename it

tadejsv commented:
For some reason tests on Ubuntu in CI complete testing are hanging at [one particular test](#https://github.com/PyTorchLightning/pytorch-lightning/pull/4838/checks?check_run_id=1514409801). I suspect it has something to do with sync_dist stuff, but I am not sure. On my local machine, also Ubuntu, this doesn't happen.

Can anyone help me fix this, or tell me how I would go about debugging it?

Borda reviewed (approved):
lgtm

teddykoker reviewed (approved):
Awesome work :)

SkafteNicki reviewed (approved):
great job @tadejsv :]

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,55 @@
+    For multi-label and multi-dimensional multi-class inputs, this metric computes the "global"
+    accuracy by default, which counts all labels or sub-samples separately. This can be
+    changed to subset accuracy (which requires all labels or sub-samples in the sample to
+    be correctly predicted) by setting `subset_accuracy=True`.
```
```suggestion
    be correctly predicted) by setting ``subset_accuracy=True``.
```

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -82,24 +113,35 @@ def __init__(
         self.add_state("correct", default=torch.tensor(0), dist_reduce_fx="sum")
         self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
 
+        if not 0 <= threshold <= 1:
+            raise ValueError("The `threshold` should lie in the [0,1] interval.")
         self.threshold = threshold
 
+        if top_k is not None and top_k <= 0:
+            raise ValueError("The `top_k` should be an integer larger than 1.")
+        self.top_k = top_k
+
+        self.subset_accuracy = subset_accuracy
```
```suggestion
        if not 0 <= threshold <= 1:
            raise ValueError("The `threshold` should lie in the [0,1] interval.")

        if top_k is not None and top_k <= 0:
            raise ValueError("The `top_k` should be an integer larger than 1.")

        self.threshold = threshold
        self.top_k = top_k
        self.subset_accuracy = subset_accuracy
```

rohitgr7 replied:
is there a check for topk > num_classes??

tadejsv replied:
Yes, this check is in the formatting function. In fact it checks topk >= num_classes, as topk = num_classes is completely useless for all purposes

rohitgr7 replied:
awesome :+1:

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -82,24 +113,35 @@ def __init__(
         Args:
-            preds: Predictions from model
+            preds: Predictions from model (probabilities, or labels)
             target: Ground truth values
```
```suggestion
            target: Ground truth labels
```

rohitgr7 commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+    Args:
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            `(0,1)` predictions, in the case of binary or multi-label inputs. Default: `0.5`
```
```suggestion
            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```

rohitgr7 commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When `None`, DDP
```
```suggestion
            Callback that performs the allgather operation on the metric state. When ``None``, DDP
```

rohitgr7 reviewed (approved):
Nice work. Just some comments.

rohitgr7 commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,120 @@
+
+    Args:
+        preds: Predictions from model (probabilities, or labels)
+        target: Ground truth values
```
```suggestion
        target: Ground truth labels
```

rohitgr7 commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+
+        Args:
+            preds: Predictions from model (probabilities, or labels)
+            target: Ground truth values
```
```suggestion
            target: Ground truth labels
```

rohitgr7 commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+    Args:
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: `0.5`
```
this actually looks super weird in the docs with ``

tadejsv replied:
Sorry, I thought you added this. Will remove.

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,55 @@
         threshold:
-            Threshold value for binary or multi-label logits. default: 0.5
+            Threshold probability value for transforming probability predictions to binary
+            `(0,1)` predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```

SkafteNicki commented:
@tadejsv have you figured out why some of the tests are hanging?

tadejsv commented:
> @tadejsv have you figured out why some of the tests are hanging?

Unfortunately not

tadejsv commented:
@SkafteNicki Figured out what is causing the tests to hang: when `ddp=True` the tests will hang when `torch.topk` is called (from the formatting function) from `run_class_metric_test`. Do you have any idea why this is happening and how to solve it? 
@Borda @teddykoker @justusschock @rohitgr7

Borda commented:
@tchaton @SeanNaren ^^ DDP

tadejsv commented:
@Borda roablock removed, tests are passing - we are ready to merge :)

I just changed `torch.topk` for `torch.argsort` in `select_topk`. This fixes the hanging tests, but I still think it should be investigated why `torch.topk` causes them to hang. I think the issue is not in DDP but in `multiprocessing.pool` - when I replace `pool.starmap` with `multiprocessing.spawn` hanging didn't happen - although then the tests all failed, seems that ddp was then not set up correctly. But anyways, this is beyond the scope of this PR :)

Borda commented:
@tadejsv sorry for the releasing inconvenience, mind rebase on the `1.2-dev` branch?

tadejsv commented:
Sure, no problem

On Wed, Dec 16, 2020 at 11:55 AM Jirka Borovec <notifications@github.com>
wrote:

> @tadejsv <https://github.com/tadejsv> sorry for the releasing
> inconvenience, mind rebase on the 1.2-dev branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PyTorchLightning/pytorch-lightning/pull/4838#issuecomment-746095408>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACXVD3BX7Q6X6HIIN4XHB6DSVCG2RANCNFSM4UBNG7JA>
> .
>

tadejsv commented:
@Borda any chance we can get this merged :) ?

Borda commented:
> @Borda any chance we can get this merged :) ?

yes, it ready, let me check why it cannot be merged...
@justusschock @teddykoker mind re-review? (your approvals was reseted by the system)

justusschock commented on notebooks/04-transformers-text-classification.ipynb:
```diff
@@ -1,5 +1,12 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
```
Why is this included in this PR?

tadejsv replied:
That was a mistake, probably bc I rebased to master at some point (after dev1.2 was created). Fixed now

justusschock commented on notebooks/05-trainer-flags-overview.ipynb:
```diff
@@ -1,5 +1,12 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
```
Also shouldn't be part of this pr

Borda commented:
@tadejsv mind rebase on `release/1.2-dev` to remove all the wrong edits?

SkafteNicki commented:
@Borda approvals still seems to be dismissed by new commits

tadejsv commented:
@Borda I already rebased (and I did it again right now). The problem is that I rebased to master in the time period between when 1.2-dev branch was created, and when it was announced in the slack channel. In my opinion that branch should have been rebased to master right before it was announced.

It doesn't contain, for example, PR #5134, which is "accidentally" picked up in this PR. If I remove this then tests here will fail, if I disabled DDP tests then another PR will be needed once 1.2-dev is rebased on master to re-enable these tests...

Borda commented:
> @Borda approvals still seems to be dismissed by new commits

I know, I am pushing @williamFalcon as I can... @edenlightning @tchaton

Borda commented:
@tadejsv it seems we resolved the issue with dismissed approvals, but mind resolve conflicts with `release/1.2-dev` so we can go... 🐰 
@justusschock @teddykoker @ananthsub @SkafteNicki mind have a quick look, I believe this was once already approved :]

tadejsv commented:
@Borda conflicts resolved, we're ready to go :)

Borda commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+    Args:
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            `(0,1)` predictions, in the case of binary or multi-label inputs.
```

Borda commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
```
```suggestion
            Forward only calls ``update()`` and return None if this is set to False.
```

Borda commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
```
```suggestion
            before returning the value at the step.
```

Borda commented on pytorch_lightning/metrics/classification/hamming_distance.py:
```diff
@@ -0,0 +1,105 @@
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When ``None``, DDP
+            will be used to perform the allgather. default: None
```
```suggestion
            will be used to perform the all gather.
```

Borda commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,55 @@
+            Should be left at default (``None``) for all other types of inputs.
+        subset_accuracy:
+            Whether to compute subset accuracy for multi-label and multi-dimensional
+            multi-class inputs (has no effect for other input types). Default: `False`
```
```suggestion
            multi-class inputs (has no effect for other input types).
```

Borda commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,33 +16,55 @@
+            ``preds = preds.flatten()`` and same for ``target``). Note that the ``top_k`` parameter
+            still applies in both cases, if set.
         compute_on_step:
             Forward only calls ``update()`` and return None if this is set to False. default: True
```
```suggestion
            Forward only calls ``update()`` and return None if this is set to False.
```

Borda commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,120 @@
+        target: Ground truth labels
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            `(0,1)` predictions, in the case of binary or multi-label inputs.
```

Borda commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,120 @@
+            Should be left at default (``None``) for all other types of inputs.
+        subset_accuracy:
+            Whether to compute subset accuracy for multi-label and multi-dimensional
+            multi-class inputs (has no effect for other input types). Default: `False`
```
```suggestion
            multi-class inputs (has no effect for other input types).
```

Borda commented on pytorch_lightning/metrics/functional/hamming_distance.py:
```diff
@@ -0,0 +1,69 @@
+    Args:
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
```
```suggestion
            (0,1) predictions, in the case of binary or multi-label inputs.
```

Borda commented on pytorch_lightning/metrics/functional/hamming_distance.py:
```diff
@@ -0,0 +1,69 @@
+
+    Accepts all input types listed in :ref:`metrics:Input types`.
+
+    Args:
```
```suggestion
    Args:
        preds: Predictions from model
        target: Ground truth
```

Borda reviewed (approved):
lgtm

rohitgr7 commented on pytorch_lightning/metrics/classification/accuracy.py:
```diff
@@ -16,35 +16,57 @@
+    Accepts all input types listed in :ref:`metrics:Input types`.
 
     Args:
         threshold:
```
just wondering, threshold be None too by default, because it's not used when we provider pred_labels, and use 0.5 by default when we have prod_probs

tadejsv replied:
Well, I guess this could be done - but in a way that if pred_probs are passed in, None would default to 0.5. otherwise this would be a very disturbing breaking change for people used to using accuracy without extra params

rohitgr7 replied:
yeah, that's what meant here `0.5 by default when we have prod_probs`

tadejsv replied:
Got it, will add this in the next PR.

rohitgr7 reviewed (approved):
Just one comment. Else LGTM. Nice work!!

tchaton commented on pytorch_lightning/metrics/functional/accuracy.py:
```diff
@@ -0,0 +1,120 @@
+        tensor(0.6667)
+    """
+
+    if top_k is not None and top_k <= 0:
```
We should test top_k is not a float.

tchaton commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -229,47 +229,6 @@ def stat_scores_multiple_classes(
-    return class_reduce(tps, sups, sups, class_reduction=class_reduction)
-
-
 def _confmat_normalize(cm):
```
Should we provide the normalisation dimension ?

tadejsv replied:
For what?

tchaton replied:
Some people like to visualise both matrices.

tadejsv replied:
Sorry, I don't understand. If you are speaking about the confusion matrix - that is not part of this PR (this one is about the accuracy metric).

tchaton reviewed (commented):
Looks great ! minor comments

SkafteNicki reviewed (approved):
LGTM, wait with merge to after release branch gets synced with master tomorrow. Then this PR just needs to updated such that the fix from #5134 is not part of it anymore.

tchaton commented:
> LGTM, wait with merge to after release branch gets synced with master tomorrow. Then this PR just needs to updated such that the fix from #5134 is not part of it anymore.

Oh sorry @SkafteNicki, I didn't have your message when I merged. I refreshed the page and saw it.

SkafteNicki commented:
@tchaton no probs. What should we do about it? Solve merge conflicts when `master` is merged into `release/1.2-dev` or revert this PR?
